### ISSUE TITLE
Documentation lists non-working Promise example Actions

### ISSUE THREAD

**Issue by abritinthebay** [goatslacker/alt#584]
this code: 
````alt.createActions({
  fetchUser(id) {
    return function (dispatch) {
      dispatch(id);
      return http.get('/api/users/' + id)
        .then(this.fetchUserSuccess.bind(this))
        .catch(this.fetchUserFailure.bind(this));
    };
  },

  fetchUserSuccess: x => x,
  fetchUserFailure: x => x,
});

alt.actions.fetchUser(123).then(() => {
  renderToHtml(alt); // we know user fetching has completed successfully, and it's safe to render the app
});
````

Will flat out not work. What will happen is that you'll get an error thrown saying `alt.actions.fetchUser.then is not a function` which of course it isn't.

What's actually returned is the anon function that's passed dispatch. This makes using promises AND dispatching the action outright impossible right now (as far as I can tell from the documentation anyhow). It appears that `this.dispatch` was removed (also spread around in the examples and docs) so there's no way to dispatch the action if you return a promise correctly.

ie - this works for `alt.actions.fetchUser(123).then(()`:

````
fetchUser(id) {
    return http.get('/api/users/' + id)
        .then(this.fetchUserSuccess.bind(this))
        .catch(this.fetchUserFailure.bind(this));
  }
````

But the action will never be dispatched. This is doubly troublesome for Server Side rendering. 

Thoughts? Currently it seems that promise support has been kind of butchered. I'm assuming it's due to some roadmap changes...

**Comment by abritinthebay** [comment]
To give you the example I'd LOVE to be able to do with this...

Right now I do not appear to have a way of saying "load this enpoint THEN pass the data into a Store THEN let me know it's safe to take a snapshot.

Obviously this is needed for server side rendering and snapshot/bootstrap functionality for the front-end.

**Comment by abritinthebay** [comment]
If it's intended to not have `this.dispatch` available (and that makes sense) I would _strongly_ suggest changing this code in `src/actions/index.js`:
````
    // async functions that return promises should not be dispatched
    if (result !== undefined && !isPromise(result)) {
      if (fn.isFunction(result)) {
        result(dispatch, alt)
      } else {
        dispatch(result)
      }
    }
````
to this:

````
    // async functions that return promises should not be dispatched
    if (result !== undefined) {
      if (!isPromise(result) {
        result.then((data) => {
            dispatch(data)
        });
      } else if (fn.isFunction(result)) {
        result(dispatch, alt)
      } else {
        dispatch(result)
      }
    }
````

That way promises work sanely, you can dispatch events from async actions. and you can keep the (I assume intended) currently broken `function(dispatch)` syntax for whatever else you need it for.

**Comment by sjmueller** [comment]
@abritinthebay yea I'm also pretty confused as to how dispatch and promises play nice together.  I tried to upgrade to 0.18.x today, and nothing works.  As of 0.17.9, most of my actions look something like this:
```
  async login(email, password) {
    const credentials = { email, password };
    try {
      this.dispatch(credentials);
      const user = await auth(credentials);
      this.actions.loginSuccess(user);
    } catch (error) {
      this.actions.loginFail(error);
    }
  }
```
I *thought* that in the cases where I didn't actually need dispatch (just needed to invoke actionSuccess, and actionFail), then my code could stay the same and just remove the one `this.dispatch()` line.  That doesn't work at all either.

**Comment by sjmueller** [comment]
Turns out the code above works as long as `this.dispatch' is removed.  The trick is to also refactor `this.actions.action()` to simply `this.action()`

**Comment by abritinthebay** [comment]
@sjmueller surely that means your action will never be dispatched?

It seems right now that dispatching is utterly broken. Data Sources also don't dispatch either, they call the actions but no dispatch which makes them functionally useless.

0.18 pretty much appears fundamentally busted to my eyes.

**Comment by sjmueller** [comment]
@abritinthebay, that is correct; the original action will never be dispatched.  However, most of the time my stores are not listening to the original action, but rather the actions for success and failure.  In the case where async actions need to dispatch, it can be written with a thunk (es2016):
```
  login(email, password) {
    const credentials = { email, password };
    return async (dispatch) => {
      try {
        dispatch(credentials);
        const user = await auth(credentials);
        this.loginSuccess(user);
      } catch (error) {
        this.loginFail(error);
      }
    };
  }
```
or es2015 syntax:
```
  login(email, password) {
    const credentials = { email, password };
    return (dispatch) => {
      dispatch(credentials);
      auth(credentials)
        .then(this.loginSuccess)
        .catch(this.loginFail);
    };
  }
```

For your case, all you need to do is wrap your promise chain in an arrow function as I did in the second example above.

Hope this helps!

**Comment by abritinthebay** [comment]
@sjmueller 

Yes, I've tried that. Unfortunately that doesn't work for me for two reasons:

1. The success actions are run, but not dispatched (ie - things listening to those actions are never alerted they happened). This, of course, is totally unhelpful

2. There is no way to run an action that returns a promise (the above returns a function, not a promise) so it's impossible to server-side render based on the action. ie - I can't do `SomeAction().then(() => { renderApp(); }` so the only way to do that would be create a redundant store that listens for the success actions... except.. well, see problem 1.

Right now, as far as I can tell, Alt is just fundamentally broken and unusable. I'd love a workaround but it's broken idiomatically at least.

**Comment by dthorpe** [comment]
+1 for the proposed edit to dispatch after returned promise completes

**Comment by djmccormick** [comment]
+1

**Comment by goatslacker** [comment]
Published fix in `0.18.2`
